### PR TITLE
[CUDA][MPS] Fix torch.arange bound validation for large float inputs

### LINKED ISSUE #153133: Inconsistent behavior between CPU and GPU implementations of `torch.arange`
## Description

When using `torch.arange()` with a start value greater than the end value and a positive step, the behavior differs between CPU and GPU implementations:
- GPU silently returns an empty tensor
- CPU correctly raises an exception about inconsistent bounds with step sign

## Reproduction Steps

```python
import torch
print(torch.__version__)

# This should fail on both CPU and GPU since it's an impossible range
try:
  print("GPU Results:", torch.arange(start=1549556900, end=1549556828, step=1989724, dtype=torch.float, device='cuda'))
except Exception as e:
  print("GPU Exception:", e)

try:
  print("CPU Results:", torch.arange(start=1549556900, end=1549556828, step=1989724, dtype=torch.float, device='cpu'))
except Exception as e:
  print("CPU Exception:", e)
```

## Actual Behavior
```
2.6.0+cu124
GPU Results: tensor([], device='cuda:0')  # Silently returns empty tensor
CPU Exception: upper bound and larger bound inconsistent with step sign  # Raises exception
```

## Expected Behavior

Both implementations should raise the same exception.

### Versions

Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 17.0.6 (++20231209124227+6009708b4367-1~exp1~20231209124336.77)
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 570.133.20
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        39 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i9-13900F
CPU family:                           6
Model:                                183
Thread(s) per core:                   2
Core(s) per socket:                   24
Socket(s):                            1
Stepping:                             1
CPU max MHz:                          5600.0000
CPU min MHz:                          800.0000
BogoMIPS:                             3993.60
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            896 KiB (24 instances)
L1i cache:                            1.3 MiB (24 instances)
L2 cache:                             32 MiB (12 instances)
L3 cache:                             36 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] triton==3.2.0
[conda] Could not collect

cc @malfet

### PR DESCRIPTION
**Pull request by narekmalk**
Fixes #153133

Fixes an inconsistency in `torch.arange` on CUDA and MPS backends when using `float32` and large input values. Previously, invalid ranges (e.g., start > end with a positive step) could silently return empty tensors due to precision loss in validation logic.

The fix introduces double precision validation for checking whether the step sign is consistent with the range direction.

This ensures `torch.arange` behaves consistently with CPU for large `float32` inputs, and raises an appropriate error when the range is invalid.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by narekmalk** [comment]
@eqy @syed-ahmed @kulinseth @malfet please take a look when you have time

File: test/test_tensor_creation_ops.py [hunk 1]
```diff
@@ -2536,6 +2536,13 @@ def test_arange_inference(self, device):
                                    torch.tensor(3),
                                    torch.tensor(1, dtype=torch.int16)).dtype)
 
+    def test_arange_float_inconsistent_bounds(self):
```
Comment (malfet) [inline test/test_tensor_creation_ops.py:2539]: Please just add the case in question to https://github.com/pytorch/pytorch/blob/365ac49840105918c604a6b1c7e81c1ca59e37fb/torch/testing/_internal/common_methods_invocations.py#L778-L782

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for working on the change.
But I think one needs to modify it a bit more before landing:
- Instead of repeating the same check in 4 places, please add something like
`inline void arange_check_bounds(const c10::Scalar& start, const c10::Scalar& end, const c10::Scalar& step)` method to RangeUtils.h and call it from the 4 places you've discovered
- Instead of adding new test, just add error input to https://github.com/pytorch/pytorch/blob/365ac49840105918c604a6b1c7e81c1ca59e37fb/torch/testing/_internal/common_methods_invocations.py#L778

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -776,11 +776,10 @@ def sample_inputs_add_sub(op, device, dtype, requires_grad, **kwargs):
         yield SampleInput(lhs, args=(rhs,), kwargs={'alpha': False})
 
 def error_inputs_arange(op, device, **kwargs):
-    yield ErrorInput(SampleInput(0, args=(3, 0)), error_type=RuntimeError, error_regex='step must be nonzer')
-    yield ErrorInput(SampleInput(0, args=(-3, 2)), error_type=RuntimeError, error_regex='bound inconsistent with step sign')
-    yield ErrorInput(SampleInput(0, args=(3, -2)), error_type=RuntimeError, error_regex='bound inconsistent with step sign')
+    yield ErrorInput(SampleInput(0, args=(3, 0, 0)), error_type=RuntimeError, error_regex='step must be nonzero')
+    yield ErrorInput(SampleInput(0, args=(3, -2)), error_type=RuntimeError, error_regex='upper bound and lower bound inconsistent with step sign')
+    yield ErrorInput(SampleInput(0, args=(1549556900, 1549556828, 1989724)), error_type=RuntimeError, error_regex='upper bound and lower bound inconsistent with step sign')
```
Comment (narekmalk) [inline torch/testing/_internal/common_methods_invocations.py:781]: @malfet, the test cases seemed to be outdated, so I updated all of them, please check

**Comment by narekmalk** [comment]
@malfet, I have addressed your comments
